Switch to GH actions for testing

ehelms opened the pull request:
This is my first time writing GH actions by hand and I have based this on other projects in our community thus this will likely need some iteration to get right. I have already pushed empty files to allow iterating on the results of the tests.

This is motivated by Travis becoming increasingly slow and GH actions providing more concurrency. We also should not use Jenkins here given this repository bootstrap's Jenkins.

evgeni commented on .github/workflows/puppet.yml:
```diff
@@ -0,0 +1,47 @@
+      - uses: actions/cache@v2
+        with:
+          path: vendor/bundle
+          key: ${{ runner.os }}-${{ matrix.ruby }}-gems-unit-${{ hashFiles('**/Gemfile.lock') }}
```
we don't have a `Gemfile.lock` in the repo, so this cache will never hit. See https://github.com/theforeman/foreman-installer-modulesync/pull/106#issuecomment-718022711 for a similar discussion.

evgeni commented on .github/workflows/puppet.yml:
```diff
@@ -0,0 +1,47 @@
+          ./g10k -quiet -cachedir .g10k/cache -puppetfile -puppetfilelocation puppet/Puppetfile_forge -moduledir puppet/forge_modules
+
+          cd puppet
+          bundle exec rake syntax
```
```
---> syntax:hiera:yaml
ERROR: Failed to parse vendor/bundle/ruby/2.4.0/gems/puppet-6.19.1/spec/fixtures/unit/data_providers/environments/hiera_bad_syntax_yaml/data/bad.yaml: (vendor/bundle/ruby/2.4.0/gems/puppet-6.19.1/spec/fixtures/unit/data_providers/environments/hiera_bad_syntax_yaml/data/bad.yaml): found unexpected ':' while scanning a plain scalar at line 3 column 1
```

Seems it tries to parse the installed files and chokes on an explicitly bad one?

ehelms replied:
Guess I need to find a way to ignore that path or to put that bundle path somewhere else

evgeni commented on .github/workflows/puppet.yml:
```diff
@@ -0,0 +1,47 @@
+name: Unit tests
```
it feels weird to have two files, both having `name: Unit tests`, and one `unit` job, but doing different things.

IMHO this should be:
1. one file called `main.yml` or `tests.yml` or whatever
2. having a useful name (or just "CI")
3. having two jobs defined: `puppet` and `jenkins`.

ehelms commented:
Switched to a `main.yml` with two sections. Applied the cache strategy @evgeni linked and fixed the errors by placing the bundle in `~/.vendor` as well as it checks out the submodules now for puppet.

evgeni commented on .github/workflows/main.yml:
```diff
@@ -0,0 +1,92 @@
+          pip install requests
+      - name: Run tests
+        run: |
+          for d in puppet/modules/jenkins_job_builder/files/theforeman.org; do
+            DIR="."
+            ( cd $d && jenkins-jobs -l debug test -r $DIR )
+            ( cd $d && jenkins-jobs test -r $DIR --config-xml -o output && find output -name config.xml -exec python ../jenkins-lint.py --xml {} + && rm -rf output )
+          done
+
+          for d in puppet/modules/jenkins_job_builder/files/centos.org; do
+            DIR="jobs"
+            ( cd $d && jenkins-jobs -l debug test -r $DIR )
+            ( cd $d && jenkins-jobs test -r $DIR --config-xml -o output && find output -name config.xml -exec python ../jenkins-lint.py --xml {} + && rm -rf output )
+          done
```
the `for d in <path>` blocks seem rather unnecessary as they only contain one item each, but then again the two could be easily folded together if we could set `DIR` more nicely.

shouldn't block this PR tho, as I think that part is copied from travis.

ehelms replied:
Yes, taken straight from Travis. So I like re-factoring afterwards rather than before.

evgeni commented on .github/workflows/main.yml:
```diff
@@ -0,0 +1,92 @@
+          - "2.4"
+    name: Puppet
+    steps:
+      - uses: actions/checkout@v2
+      - name: Checkout submodules
+        uses: textbook/git-checkout-submodule-action@master
```
```suggestion
      - uses: actions/checkout@v2
        with:
          submodules: true
```

ekohl commented on .github/workflows/main.yml:
```diff
@@ -0,0 +1,92 @@
+          gem install bundler
+          cd puppet
+          bundle config path ~/.vendor/bundle
+          bundle lock
```
Locking wasn't needed in my experience.

ehelms replied:
How did you create the lock file to be used in the cache step without this?

ekohl replied:
You don't? https://github.com/theforeman/foreman-installer-modulesync/pull/106#issuecomment-720644253 explains why it's not needed.

ekohl commented on .github/workflows/main.yml:
```diff
@@ -0,0 +1,92 @@
+          bundle install --jobs $(nproc) --retry 3
+      - name: Run tests
+        run: |
+          if [[ ! -f g10k ]] ; then
+            if [[ ! -f g10k-linux-amd64 ]] ; then
+              wget https://github.com/xorpaul/g10k/releases/download/v0.8.9/g10k-linux-amd64.zip
+            fi
+            unzip g10k-linux-amd64.zip
+          fi
+
+          ./g10k -quiet -cachedir .g10k/cache -puppetfile -puppetfilelocation puppet/Puppetfile_forge -moduledir puppet/forge_modules
```
Might be good to do this as a separate step since it's not really part of running tests, really just installing modules. Also, might be nice to cache `.g10k/cache`

ekohl commented on .github/workflows/main.yml:
```diff
@@ -0,0 +1,92 @@
+      fail-fast: false
+      matrix:
+        ruby:
+          - "2.4"
```
Puppet 6 includes Ruby 2.5.
```suggestion
          - "2.5"
```

It's also confusing to read `Puppet (2.4)` in the checks.

ekohl commented on .github/workflows/main.yml:
```diff
@@ -0,0 +1,86 @@
+        uses: actions/setup-python@v1
+        with:
+          python-version: '2.7'
+      - name: Restore pip cache
+        uses: actions/cache@v2
+        with:
+          path: ~/.cache/pip
+          key: ${{ runner.os }}-pip-${{ hashFiles('**/requirements*.txt') }}
+          restore-keys: |
+            ${{ runner.os }}-pip-
```
We don't have a `requirements.txt`. I think that means the cache will never be rebuilt.

@evgeni would a solution to be to add a `pip freeze > requirements-lock.txt` at the end of the install phase?

evgeni replied:
Technically? Yeah.

But we also could just dump the three packages on the next step into a `requirements.txt` and use that? That would produce a rather static hash, as the non-versioned deps don't change much, but I think that's still OK?

ekohl commented on .github/workflows/main.yml:
```diff
@@ -0,0 +1,86 @@
+      - name: Setup ruby
+        uses: actions/setup-ruby@v1
+        with:
+          ruby-version: '2.6'
```
2.6 is an odd version since there's no Puppet version that ships it. Probably doesn't matter too much, but using 2.5 ensures you don't rely on any Ruby 2.6 features.

ekohl commented on .github/workflows/main.yml:
```diff
@@ -0,0 +1,86 @@
+          done
+  puppet:
+    runs-on: ubuntu-latest
+    strategy:
+      fail-fast: false
```
Fail fast is only relevant with a matrix build, but it's no longer a matrix.

ekohl commented on .github/workflows/main.yml:
```diff
@@ -0,0 +1,86 @@
+      - uses: actions/cache@v2
+        with:
+          path: ~/.vendor/bundle
+          key: ${{ runner.os }}-${{ matrix.ruby }}-gems-unit-
+          restore-keys: |
+            ${{ runner.os }}-${{ matrix.ruby }}-gems-unit-
```
There is no more matrix.

ekohl commented on .github/workflows/main.yml:
```diff
@@ -0,0 +1,86 @@
+jobs:
+  jjb:
+    runs-on: ubuntu-latest
+    strategy:
+      fail-fast: false
```
This is only relevant for a matrix

ekohl commented on .github/workflows/main.yml:
```diff
@@ -0,0 +1,86 @@
+          bundle install --jobs $(nproc) --retry 3
+      - name: Setup g10k
+        run: |
+          if [[ ! -f g10k ]] ; then
+            if [[ ! -f g10k-linux-amd64 ]] ; then
+              wget https://github.com/xorpaul/g10k/releases/download/v0.8.9/g10k-linux-amd64.zip
+            fi
+            unzip g10k-linux-amd64.zip
+          fi
```
We don't store this in git so it's safe to simplify it:
```suggestion
          wget https://github.com/xorpaul/g10k/releases/download/v0.8.9/g10k-linux-amd64.zip
          unzip g10k-linux-amd64.zip
```

For the Travis scripts I made it idempotent because I ran them locally as well.

evgeni commented on .github/workflows/main.yml:
```diff
@@ -0,0 +1,78 @@
+    steps:
+      - uses: actions/checkout@v2
+      - name: Setup python
+        uses: actions/setup-python@v1
```
```suggestion
        uses: actions/setup-python@v2
```

can you add a dependabot config like https://github.com/theforeman/obal/blob/master/.github/dependabot.yml so that actions get bumped automatically?

evgeni commented on .github/workflows/main.yml:
```diff
@@ -0,0 +1,78 @@
+      - uses: actions/cache@v2
+        with:
+          path: ~/.vendor/bundle
+          key: ${{ runner.os }}-gems-unit-
+          restore-keys: |
+            ${{ runner.os }}-gems-unit-
```
this seems redundant, no `hashFiles()` at all here? then the cache will always be saved as `Linux-gems-unit-`

evgeni commented on .github/workflows/main.yml:
```diff
@@ -0,0 +1,78 @@
+        run: |
+          sudo apt-get update && sudo apt-get install -y yajl-tools
+          gem install bundler
+          cd puppet
```
you can set a `defaults.run.working-directory` to `puppet` so that you don't need to `cd` all the time:

https://docs.github.com/en/free-pro-team@latest/actions/reference/workflow-syntax-for-github-actions#jobsjob_iddefaultsrun

evgeni commented on .github/workflows/main.yml:
```diff
@@ -0,0 +1,78 @@
+      - name: Run tests
+        run: |
+          cd puppet
+          bundle exec rake syntax
+          bundle exec rake lint
+          bundle exec rake spec
```
would it make sense to have these as separate steps, so that it's easier to see which step failed?

ekohl commented on .github/workflows/main.yml:
```diff
@@ -0,0 +1,83 @@
+      - name: Setup python
+        uses: actions/setup-python@v2
+        with:
+          python-version: '2.7'
```
I think we should use Python 3.

ekohl commented on .github/workflows/main.yml:
```diff
@@ -0,0 +1,83 @@
+            ${{ runner.os }}-pip-
+      - name: Install dependencies
+        run: |
+          echo 'jenkins-job-builder' >> requirements.txt
+          echo 'lxml' >> requirements.txt
+          echo 'requests' >> requirements.txt
```
Can't we store `requirements.txt` in git as a file?

ehelms replied:
We could. @evgeni suggested this method, y'all figure it out and tell me what you want.

evgeni replied:
I did not, I meant to have the file in git. ;)

ekohl commented on .github/workflows/main.yml:
```diff
@@ -0,0 +1,83 @@
+        uses: actions/cache@v2
+        with:
+          path: ~/.cache/pip
+          key: ${{ runner.os }}-pip-${{ hashFiles('**/requirements*.txt') }}
```
Can you include the Python version in the cache key? That way the cache is automatically invalidated when we switch the Python version.

evgeni replied:
The cached wheels are version agnostic, so why bother?

ekohl replied:
I suspected lxml was version dependent. It's not?

evgeni replied:
lxml is, indeed, but then the filename in the cache is different too and it'll be updated from the internet. no idea what's saving more, to be honest.

ekohl replied:
I suspect that if you use a common cache, it can grow and grow. Not sure if there's anything that purges old wheels from the cache, but switching Python version could be a good enough thing to force that.

ekohl commented on .github/workflows/main.yml:
```diff
@@ -0,0 +1,83 @@
+      - uses: actions/cache@v2
+        with:
+          path: ~/.vendor/bundle
+          key: ${{ runner.os }}-gems-unit-
```
This should include the Ruby version and hashfiles
```suggestion
          key: ${{ runner.os }}-gems-2.5-${{ hashFiles('**/Gemfile.lock') }}
```

ehelms replied:
If there is never a `Gemfile.lock` at this point in the steps, I do not get why I would include hashing it?

ekohl replied:
I think that `bundle install` generates a `Gemfile.lock` and in the post step it will create a bundle with the correct hash.

ehelms replied:
That makes sense. When it runs the next time, since there is no `Gemfile.lock` created yet at this step, it will use `${{ runner.os }}-gems-2.5-` as the lookup key right? So the hash of the `Gemfile.lock` seems to be pointless without creating the lockfile in the step above it.

ekohl commented on .github/workflows/main.yml:
```diff
@@ -0,0 +1,83 @@
+          restore-keys: |
+            ${{ runner.os }}-gems-unit-
+      - name: Bundle install
+        run: |
+          bundle install --jobs $(nproc) --retry 3
```
If it's a single command, there's no need for the multi-line syntax
```suggestion
        run: bundle install --jobs $(nproc) --retry 3
```

ehelms replied:
Truth. I find it nice to read though.

tbrisker commented:
What's the status here? i'd love to get rid of travis, it's holding up a release yet again...

ekohl reviewed (approved):
Going to merge this now to get rid of Travis. If it's incomplete, we can fix it later.
